libfwupdplugin/fu-efi-firmware-volume.c

```diff
@@ -0,0 +1,386 @@
+
+#include "config.h"
+
+#include "fu-common.h"
+#include "fu-efi-common.h"
+#include "fu-efi-image.h"
+#include "fu-efi-firmware-volume.h"
+
+/**
+ * SECTION:fu-uefi-firmware-volume
+ * @short_description: UEFI Volume
+ *
+ * An object that represents a UEFI file volume image.
+ *
+ * See also: #FuFirmware
+ */
+
+G_DEFINE_TYPE (FuEfiFirmwareVolume, fu_efi_firmware_volume, FU_TYPE_FIRMWARE)
+
+#define FU_EFI_FIT_SIGNATURE					0x5449465F
+
+#define FU_EFI_FIRMWARE_VOLUME_SIGNATURE			0x4856465F
+#define FU_EFI_FIRMWARE_VOLUME_REVISION				0x02
+
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR		0x00
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_GUID			0x10
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_LENGTH			0x20
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_SIGNATURE			0x28
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_ATTRS			0x2C
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_HDR_LEN			0x30
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_CHECKSUM			0x32
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_EXT_HDR			0x34
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_RESERVED			0x36
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_REVISION			0x37
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_BLOCK_MAP			0x38
+#define FU_EFI_FIRMWARE_VOLUME_SIZE				0x40
+
+static gboolean
+fu_efi_firmware_volume_parse_one (FuEfiFirmwareVolume *self,
+				  GBytes *fw,
+				  gsize *offset_val,
+				  guint32 img_cnt,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	fwupd_guid_t guid = { 0x0 };
+	gsize blockmap_sz = 0;
+	gsize bufsz = 0;
+	gsize offset = *offset_val;
+	guint16 checksum = 0;
+	guint16 ext_hdr = 0;
+	guint16 hdr_length = 0;
+	guint32 attrs = 0;
+	guint32 sig = 0;
+	guint64 fv_length = 0;
+	guint8 revision = 0;
+	const guint8 *buf = g_bytes_get_data (fw, &bufsz);
+	g_autofree gchar *guid_str = NULL;
+	g_autoptr(FuFirmwareImage) img = NULL;
+	g_autoptr(GBytes) blob = NULL;
+
+	/* check if this is _FIT */
+	if (!fu_common_read_uint32_safe (buf, bufsz, offset,
+					 &sig, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read start signature: ");
+		return FALSE;
+	}
+	if (sig == FU_EFI_FIT_SIGNATURE) {
+		*offset_val += 0x150000;
+		return TRUE;
+	}
+	if (!fu_common_read_uint32_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_SIGNATURE,
+					 &sig, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read signature: ");
+		return FALSE;
+	}
+	if (sig != FU_EFI_FIRMWARE_VOLUME_SIGNATURE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "signature invalid, got 0x%x, expected 0x%x",
+			     sig, (guint) FU_EFI_FIRMWARE_VOLUME_SIGNATURE);
+		return FALSE;
+	}
+
+	/* check reserved section */
+	for (guint i = 0; i < 0x10; i++) {
+		if (buf[offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i] != 0x0) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "reserved section invalid @0x%x",
+				     (guint) offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i);
+			return FALSE;
+		}
+	}
```

Comment:
does the spec actually say that it should be zero?  Usually reserved means don't care - it should be zero, but a future version of a spec can encroach into this area and so looking for zero might mean a failure later.

Reply:
You're completely correct; Apple firmware seems to put some kind of CRC here. I'll remove this check.